### ISSUE TITLE
New RPC - REST - V2

### ISSUE THREAD

**Issue by jorgecuesta** [pokt-network/pocket-core#1455]
**Please explain your change in detail.**

Create a new routes under /v2 for:

* Every existent endpoint will be replicated with:
  * structured and normalized params
  * structured and normalized response
  * standard use for the response codes (200/400/404)
* Add a way to always query "latest" information and return on the response which was that height.
* Add "circulating" to supply (preferred) or a dedicated endpoint
* Add Unconfirmed Txs and Tx (both replicated assuming the Unconfirmed Tx proposal is approved and deployed)
* Rewards calculation (like relays_counter under an endpoint)
* Enhanced /sim/client with a better error handling to allow a better understanding of the errors in between pocket node -> chain
* Provide a way to return back just a few properties from the element, reducing response size on some queries that u may doesn't need the full response.
* Add an endpoint to check how many relays the node currently process and storage to be claim on the session end.
* Unit test for every single endpoint.

**Request Sample/Idea:**
```
{
  // optional - default => latest
  "height": 60000,

  "filter": {
    // nodes, apps, accounts, nodeclaims, relays, blocktxs, sessionRelays
    "addresses": ["", ""],
    "status": "",

    // only nodes - apps maybe in future?
    "jailed": true,

    // blocktxs - unconfirmedtxs (decode and search)
    "tx_hash": "",
    "tx_code": 0,
    "tx_type": "claim/proof",

    // nodes, apps, nodeclaims, relays, blocktxs,
    "chains": ["", ""]
  },

  "page_info": {
    "page": 1,
    "limit": 10
  }
}
```

**Response Sample/Idea:**
```
{
  // success
  "success":  true,
  "message": "SUCCESS MESSAGE (OK)",
  
  // error
  "success": false,
  "message": "E100000 - ERROR MESSAGE",
  
  // always
  "height": 50000,
  
  // single element
  "data": {
    "block": {},
    "block_meta": {}
  },

  // list of elements
  "items": [{}, {}],
  
  // pagination
  "page_info": {
    "page": 1,
    "limit": 10,
    "total_items": 1000,
    "has_more": true
  }
}
```

**Please provide a justification for your change.**

* Current RPC of Pocket lack on a formal params/response/codes structure.
* Many of the request that a user normally does could be done without need to first query height to later sent it.
* Many of the request can easily be done with GET instead of POST, but the plan is support both.
* Many times we use the RPC to get just a few properties.
* Many people lookup crazy ways to calculate their rewards on a period of time, other of course use POKTscan/pokt.watch or something similar, but a way to the user to be sure and check this on their own node is always welcome.
* Check relays processed for the next claim allow to quickly user check if the node is "working" and avoid restarts or stop before they are claimed.
* Simulate relays need to be better to provide a better debug layer for the user that is trying to figure out what is going wrong between pocket node and chain node. Current implementation say nothing about the error, or even if the payload for the sim is wrong.

Due to the nature of this change and because is so big, I understand that avoid to touch so drastically the /v1 is a better option to keep those node runner tools working without break anything. This could be adopted by each one on their own timelines.

**Timeline for implementation**

This is a long, long task, based on my current skill set I think this probably took to me alone a month of development. This will need a good bunch of discussion and details before apply 1 line of code.

**Comment by blockjoe** [comment]
I want to make sure this doesn't get lost, mainly because the scope of it right now is pretty broad. I want to go through some of these and see if we can start to massage a proposal(s) out of this because I think there's a lot of valuable feedback here. 

These are some initial thoughts about the initial list of points:


> - Every existent endpoint will be replicated with:
       - structured and normalized params
       
While I see how this could improve the developer experience in the sense that it's only one set of commands to remember, the trade-off is that it makes it harder for clients to have static type checking of their implementations. While the benefit is that a single request structure is easier to remember, remembering the context of when parameters are required, optional, or unused now falls back to the client. 

I'm inclined to advocate for unique request models for each route, as it makes for a better experience with auto documentation and code generation tools, I will acknowledge that the current v1 of the RPC API hasn't kept the most usable spec in those regards. 

** I'm curious if this could be addressed by having a more accurate spec that would generate valid clients from codegen tools? **

> - structured and normalized response
      
I think this has a more direct and immediate benefits, however, with a caveat of a verbose toggle. While more information for consumer clients is always going to be preferred, it does come with some bandwidth considerations. 

On the considered fields:

- `success: bool` -> I think there is a need for improving error typing in the specification, and if this is needed to help discriminate that to a client, I'm a fan.
- `message: str` -> If the success message will always be the same, then I think we could include the message in an explicitly typed Error response, which would populate the `data` field.
- `height: int` -> If this is the height at execution time of the request, I'm a really big fan of this, and perhaps **something worth considering in a proposal immediately in v0**.
- `data: ResponseTypeVar` -> I tend to gravitate towards `result`, but agree that wrapping in it's own field like this is a better move. The `ResponseTypeVar` I'm referring to in this case would be the existing response type for that route.
- `items: list` -> I think that this could simply live in the `data/result` field be defined in the type for that route, because if not, I as a client have to figure out when exactly I should be using `items` vs `data`.
- `pagination` -> I believe this type is standardized, but again, I think is better to absorbed into the type definition for the response under `data/result`
 
Additionally though, I think we should include a `version` field similar to the height that populates the existing version of the API in use, something like `v1/0.8.3-RC`, `v1/0.9.0`, `v2/0.9.1` etc. 

 tldr of above; We need to: be **better at discriminating errors**, should probably consider having the **height in all responses already**, push the **current response into its own field**, and add a **version field** to responses.  
 
>  - standard use for the response codes (200/400/404)

From the perspective of strict HTTP(S), I agree with this. When thinking from the perspective of websockets and possibly other transport interfaces, this standardization doesn't end up completely translating back to clients. While I **don't think HTTP codes are the answer** here, I do think **we need to be typing our errors and including them in the spec** so that they can be handled properly by all clients in a standard way.

> - Add a way to always query "latest" information and return on the response which was that height.

I was pretty positive **an empty height field already allowed you to query latest information** in the current version. Returning the height in the response though is a massive improvement to this functionality though. When using Pocket Network to consume Pocket Network data 🤯, the fact that two nodes in the session could be at different blocks as they change actually makes something like calling `/query/height` and then `/query/supply` at the returned height can fail, because those heights might not match. 

**Adding the height in the response is a great addition**, and would quickly allow Pocket Network consumers reading from a session to be able to mitigate these race conditions that are going to naturally happen due to the p2p nature of the underlying blockchain network.

> - Add "circulating" to supply (preferred) or a dedicated endpoint

What would be the definition of circulating here, `TotalSupply - SumAppStakes - SumNodeStakes - DAOTreasuryBalance`? If it can be currently calculated from on-chain data, but not from the current supply response, **perhaps this is also something else worth tackling in the current API?**

> - Add Unconfirmed Txs and Tx (both replicated assuming the Unconfirmed Tx proposal is approved and deployed)

Unconfirmed is a needed endpoint, but I'm inclined to believe it should live as a `/private` route, optionally exposed. Only reason for this is that the mempool responses are not going to be deterministic across the network, and will be unique to that individual node.

> - Rewards calculation (like relays_counter under an endpoint)

This seems like it should be low-hanging fruit, but for how things are currently implemented, this is a non-trivial task. Nodes don't come to consensus about the state of proof transactions for each node and if they resulted in a payment. In addition to this, the total number of relays is only stored in the claim. It's pretty trivial right now to get one of the following two things:

- Total number of claimed relays by a node
- Total amount of POKT rewarded for proofs

If you want to calculate total number of rewarded relays, that means correlating the history of claim transactions from a node to their corresponding proof transactions if they exist, and conditionally counting the number of claims from the claim only if there is a corresponding proof. **With moving the persistence layer towards SQL, this should become significantly easier.**

> Enhanced /sim/client with a better error handling to allow a better understanding of the errors in between pocket node -> chain

I'll need to better explore what happens here, and if we're currently swallowing errors that you would see by calling the node directly. If we aren't swallowing the response of the relay chain though, unfortunately, the ambiguity of those errors should be addressed with the developers of that relay chain.

> - Provide a way to return back just a few properties from the element, reducing response size on some queries that u may doesn't need the full response.

Again, going back to that static type analysis for clients, while this has obvious benefits, I think the better course of action is identifying where this is currently happening, and if it can be addressed without conditional request/response formatting if possible.

> - Add an endpoint to check how many relays the node currently process and storage to be claim on the session end.

Also should have the `/private` context to avoid nodes being flooded offline to force them to lose unclaimed rewards.

Hoping to get some conversations started around this to see what we can take action on from these insights.
